crates/accelerate/src/synthesis/discrete_basis/solovay_kitaev.rs

```diff
@@ -0,0 +1,276 @@
+#[pyclass]
+pub struct SolovayKitaevSynthesis {
+    /// The set of basic approximations.
+    basic_approximations: BasicApproximations,
+    /// Whether to perform runtime checks on the handled matrices/data.
+    do_checks: bool,
+}
+
+impl SolovayKitaevSynthesis {
+    /// Initialize a new instance by generating a set of basic approximations, given a list
+    /// of discrete standard gates and a depth. Returns an error if any gate is not discrete.
+    pub fn new(
+        basis_gates: &[StandardGate],
+        depth: usize,
+        tol: Option<f64>,
+        do_checks: bool,
+    ) -> Result<Self, DiscreteBasisError> {
+        let basic_approximations = BasicApproximations::generate_from(basis_gates, depth, tol)?;
+        Ok(Self {
+            basic_approximations,
+            do_checks,
+        })
+    }
+
+    /// Run the Solovay Kitaev algorithm, given the initial unitary as U(2) matrix.
+    ///
+    /// This matrix is given using [Complex64] numbers, which can limit the precision of the
+    /// algorithm. It is preferred to run the algorithm using [synthesize_std], which let's Qiskit
+    /// attempt to construct the underlying SO(3) matrix representation using [BigFloat] at
+    /// quadruple precision, instead of inferring it from the double precision U(2) matrix.
+    pub fn synthesize_matrix(
+        &self,
+        matrix_u2: &Matrix2<Complex64>,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_u2(matrix_u2, self.do_checks);
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run the Solovay Kitaev algorithm on a standard gate + gate parameters.
+    ///
+    /// This attempts to directly construct the SO(3) matrix representation to minimize roundoff
+    /// errors. If unsuccessful, this falls back onto constructing the standard U(2) matrix
+    /// representation and converting it to SO(3).
+    pub fn synthesize_std(
+        &self,
+        gate: &StandardGate,
+        params: &[Param],
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_std(gate, params)?;
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run Solovay Kitaev given a [GateSequence].
+    ///
+    /// Mainly an internal helper for other runner methods.
+    fn synthesize_sequence(
+        &self,
+        sequence: &GateSequence,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        // Run SK recursion.
+        let mut output = self.recurse(sequence, recursion_degree);
+
+        // Do minor optimizations on the output sequence, e.g. cancel gate-inverse-pairs, which
+        // can arise due to the recursions (even though they should not be present in the initial
+        // set of basic approximations).
+        output.inverse_cancellation();
```

Comment:
Does this also merge two consecutive "T-gates" into an "S-gate", etc.? I think this is something we may want to do in practice, though technically speaking this would take us out from the `[T, Tdg, H]` basis-set. On the other hand, we probably do not want to consider the basis set `[T, Tdg, H, S, Sdg]`, as this would not give us any actual benefit (I think), but will make the runtime quite a bit slower (I think).